File: test/data/testUtils.py
import numpy as np


from data.testData import TestDataScalar, TestDataArray


# Obtain the index ( flat index ) from multiple indices
def obtainPos(dim, strides, indices):
    offset = 0

    if dim == 0:
        return 0

    if dim == 1:
        return indices[0]

    for i in range(0, dim):
        offset += indices[i] * strides[i]
    return offset


def obtainStrides(shape, dim):
    # Pre-allocate a tuple of zeros
    strds = [0] * dim
    # The stride for the first dimension is always 1
    strds[0] = 1

    # For subsequent dimensions, stride[i] = stride[i-1] * size[i-1]
    for i in range(1, dim):
        strds[i] = strds[i - 1] * shape[i - 1]
    return tuple(strds)

    # function pretty_index(idx...)
    #     parts = map(pretty_index, idx)
    #     return join(parts, ", ")
    # end

    # function pretty_index(idx::AbstractArray)
    #     parts = map(v ->
    #         v === Colon() ? ":" :
    #         string(v),
    #         idx
    #     )
    #     return "[" * join(parts, ", ") * "]"
    # end

    # function pretty_index(idx::Tuple)
    #     parts = map(v ->
    #         v === Colon() ? ":" :
    #         isa(v, AbstractArray) ? pretty_index(v) :
    #         string(v),
    #         idx
    #     )
    #     return "(" * join(parts, ", ") * ")"
    # end

    # function pretty_index(idx::Colon)
    #     return ":"
    # end

    # function pretty_index(idx::Number)
    #     return string(idx)
    # end




class PrettyIndex:

    @staticmethod
    def formatIndex(*idx):
        """Format indices, handling slices, tuples, numbers."""
        parts = []
        for i in idx:
            if isinstance(i, slice):
                parts.append(":")
            else:
                # Convert to 1-based index for printing
                parts.append(str(i + 1) if isinstance(i, int) else str(i))
        return "(" + ", ".join(parts) + ")"

    @staticmethod
    def format(*idx):
        """Format indices, handling slices, tuples, numbers."""
        parts = []
        for i in idx:
            if isinstance(i, slice):
                parts.append(":")
            else:
                # Convert to 1-based index for printing
                parts.append(
                    str(i + 1)
                    if isinstance(i, int) or isinstance(i, np.int64)
                    else str(i)
                )
        if len(parts) == 1:
            return parts[0]
        return ", ".join(parts)


def toJuliaCoordinates(indices):
    if len(indices) <= 1:
        return indices
    elif len(indices) == 2:
        return (indices[1], indices[0])
    else:
        tmp_array = [x for x in indices]
        end = len(tmp_array) - 1
        tmp = tmp_array[end - 1]
        tmp_array[end - 1] = tmp_array[end]
        tmp_array[end] = tmp
        return tuple(reversed(tmp_array))


def logShape_testData(testData, write_log_fn: callable):
    logShape(testData.dim, testData.total_element, testData.shape, write_log_fn)


def logShape(dim, total_element, shape, write_log_fn: callable):
    toLog = f"dim: {dim}\ntotal_element: {total_element}\nshape: {shape}"
    write_log_fn(toLog)


def log_scalar(ndarray: TestDataScalar, write_log_fn: callable):
    write_log_fn(f"{ndarray.data}")


def write_scalar_content(scalar, write_log_fn):
    ndarray = np.array(scalar.data)
    write_log_fn(f"{ndarray[()]}")
    logShape(ndarray.ndim, ndarray.size, ndarray.shape, write_log_fn)


def write_content(ndarray, write_log_fn: callable):
    if isinstance(ndarray, TestDataScalar):
        log_scalar(ndarray, write_log_fn)
    elif isinstance(ndarray, np.ndarray):
        logShape(ndarray.ndim, ndarray.size, ndarray.shape, write_log_fn)
        log_by_index(ndarray, write_log_fn)
        log_by_iterator(ndarray, write_log_fn)
        log_by_slices(ndarray.shape, ndarray, ndarray.ndim, write_log_fn)
    else:
        logShape_testData(ndarray, write_log_fn)
        log_by_index_testData(ndarray, write_log_fn)
        log_by_iterator_testData(ndarray, write_log_fn)
        log_by_slices_testData(ndarray, write_log_fn)


def log_by_index_testData(ndarray: TestDataArray, write_log_fn: callable):
    log_by_index(ndarray.matrix, write_log_fn)

def round_0(n, ndigits=7):
    if isinstance(n, float) and n == -0.0:
        n =  0.0
    elif isinstance(n, int) and n == -0:
        n = 0
    return round(n, ndigits=ndigits)

def log_by_index(ndarray, write_log_fn: callable):
    """Log each element by explicit indices, rounding to 7 digits."""
    content_by_index = ["index: "]
    matrix = np.transpose(ndarray)
    shape = matrix.shape

    for indices in np.ndindex(shape):
        element = matrix[indices]

        # Round element(s) to 7 digits
        if isinstance(element, list):
            x = [round_0(e, 7) for e in element]
        else:
            x = round_0(element, 7)

        content_by_index.append(f"{PrettyIndex.formatIndex(*reversed(indices))}: {x}, ")

    write_log_fn("".join(content_by_index))


def log_by_iterator_testData(ndarray: TestDataArray, write_log_fn: callable):
    log_by_iterator(ndarray.matrix, write_log_fn)


def log_by_iterator(matrix, write_log_fn: callable):
    """Log each element using flat iteration."""
    content_by_iterator = ["iterator: "]
    for val in matrix.flatten(order="F"):

        if isinstance(val, list):
            x = [round_0(e, 7) for e in val]
        else:
            x = round_0(val, 7)
        content_by_iterator.append(f"{x} ")
    write_log_fn("".join(content_by_iterator))


def fortran_ndindex(shape):
    """Generate indices in Fortran (column-major) order."""
    size = np.prod(shape)
    for flat in range(size):
        yield np.unravel_index(flat, shape, order="F")


def log_by_slices_testData(ndarray: TestDataArray, write_log_fn: callable):
    log_by_slices(
        ndarray.matrix.shape, ndarray.matrix, ndarray.matrix.ndim, write_log_fn
    )


def log_by_slices(shape, matrix, ndim, write_log_fn: callable):
    """Log slices along each axis, rounding all elements to 7 digits."""

    write_log_fn("slices:")
    content_by_slices = []

    for axis in range(ndim):
        # create the shape for iterating over fixed indices along other axes
        iter_shape = tuple(1 if i == axis else s for i, s in enumerate(shape))
        for idx in fortran_ndindex(iter_shape):
            # Build a tuple of slice objects for the full array
            indices = tuple(slice(None) if i == axis else idx[i] for i in range(ndim))
            slice_view = matrix[indices]

            # Flatten and round, handling elements that are lists
            rounded_elements = []
            for x in slice_view.flatten():
                if isinstance(x, (list, np.ndarray)):
                    rounded_elements.append([ round_0(e, 7) for e in x])
                else:
                    rounded_elements.append(round_0(x, 7))

            # Convert to string
            slice_str = "[" + ", ".join(str(e) for e in rounded_elements) + "]"
            content_by_slices.append(
                f"array[{PrettyIndex.format(*indices)}] = {slice_str}\n"
            )

    write_log_fn("".join(content_by_slices))

File: test/data/testUtils.jl
module TestUtils

using Algebra42

using ..TestData

export printTest, exceptionPassed, verifyShape, compareArrayContent, compareByIndex, compareByIterator, compareBySlices, exceptionPassed, allExceptionPassed, comparesSlices, write_content, logByIndex, logByIterator, logBySlices


# ---------------------------------------------------------------------------------------------------------------------------------#
#                                                                                                                                  #
#                                                  Tests Utils ( printing tests )                                                  #
#                                                                                                                                  #
# ---------------------------------------------------------------------------------------------------------------------------------#

function printTest(name::AbstractString, result::Bool)
    try
        if result === true
            println("   ✅ Test passed: $name")
        else
            println("   ❌ Test failed: $name")
        end
    catch e
        println("   ❌ Test error: $name")
        println("      → $(typeof(e)): $(e.msg)")
    end
end

function test(name::AbstractString, result::Bool)
    try
        if result === true
            println("   ✅ Test passed: $name")
        else
            println("   ❌ Test failed: $name")
        end
    catch e
        println("   ❌ Test error: $name")
        println("      → $(typeof(e)): $(e.msg)")
    end
end


function pretty_index(idx...)
    parts = map(pretty_index, idx)
    return join(parts, ", ")
end

function pretty_index(idx::AbstractArray)
    parts = map(v -> 
        v === Colon() ? ":" : 
        string(v),
        idx
    )
    return "[" * join(parts, ", ") * "]"
end

function pretty_index(idx::Tuple)
    parts = map(v -> 
        v === Colon() ? ":" : 
        isa(v, AbstractArray) ? pretty_index(v) : 
        string(v),
        idx
    )
    return "(" * join(parts, ", ") * ")"
end

function pretty_index(idx::Colon)
    return ":"
end

function pretty_index(idx::Number)
    return string(idx)
end

# ---------------------------------------------------------------------------------------------------------------------------------#
#                                                                                                                                  #
#                                                  Tests Utils ( check shape )                                                     #
#                                                                                                                                  #
# ---------------------------------------------------------------------------------------------------------------------------------#


function verifyShape(shape::Shape, expected::ShapeData)
    if shape.dims != expected.shape
        println("Shape mismatch: got $(shape.dims), expected $(expected.shape)")
    end

    if shape.length != expected.total_element
        println("Size mismatch: got $(shape.length), expected $(expected.total_element)")
    end

    if ndims(shape) != expected.dim
        println("Dim mismatch: got $(ndims(shape)), expected $(expected.dim)")
    end

    return shape.dims == expected.shape &&
           shape.length == expected.total_element &&
           ndims(shape) == expected.dim
end

# using test data
verifyShape(testData::TestDataScalar) = verifyShape(testData.data, testData)
verifyShape(testData::TestDataArray) = verifyShape(testData.array, testData)

# using data types
verifyShape(nb::Number, expected::ShapeData) = verifyShape(Shape(nb), expected)
verifyShape(tuple::Tuple, expected::ShapeData) = verifyShape(Shape(tuple), expected)
verifyShape(array::AbstractArray, expected::ShapeData) = verifyShape(Shape(array), expected)

# using ndarray
verifyShape(ndarray::NDArray, expected::ShapeData) = verifyShape(ndarray.shape, expected)

# using ragged array
verifyShape(test, expected::ShapeData, dtype) = verifyShape(Shape(test.array; dtype=dtype), expected)

function verifyShape(array1, array2)
    return size(array1) == size(array2) && ndims(array1) == ndims(array2)
end

# ---------------------------------------------------------------------------------------------------------------------------------#
#                                                                                                                                  #
#                                                  Tests Utils ( check array content )                                             #
#                                                                                                                                  #
# ---------------------------------------------------------------------------------------------------------------------------------#

function compareArrayContent(ndarray, testData::TestDataArray)
    return compareByIndex(ndarray, testData.matrix) &&
           compareByIterator(ndarray, testData.matrix) &&
           compareBySlices(ndarray, testData.matrix)
end


function compareByIndex(ndarray, expected)
    ranges = ntuple(i -> 1:size(ndarray)[i], ndims(ndarray))

    for indices in Iterators.product(ranges...)
           println("index used: ", indices)
        if (ndarray[indices...] != expected[indices...])
            println("Comparison byIndex failed at index: ", indices)
           # println(" Original was", ndarray)
           # println(" Expected was", expected)
            println("  Expected: ", expected[indices...])
            println("  Got:      ", ndarray[indices...])
            return false
        end
    end
    return true
end

function compareByIterator(ndarray, expected)
    for (got, shouldBe) in zip(ndarray, expected)
        if (got != shouldBe)
            println("Comparison byIterator failed at element ", i)
            println(" Original was", ndarray)
            println("  Expected: ", shouldBe)
            println("  Got:      ", got)
            return false
        end
    end
    return true
end

function compareBySlices(ndarray, expected)
    shape = size(ndarray)

    for axis in 1:ndims(ndarray)
        for idx in CartesianIndices(shape)

            idx_tuple = Tuple(idx)
            array = Array{Union{Int,Colon}}(undef, length(idx_tuple))

            for (i, val) in enumerate(idx_tuple)
                if (i == axis)
                    array[i] = Colon()
                else
                    array[i] = val
                end
            end

            if (!comparesSlices(ndarray, expected, array))
                return false
            end
        end

    end
    return true
end

function comparesSlices(ndarray, expected, indices)
    println("A: $ndarray")
    println("slices made: A[$(indices)]")

    slice_test::SubArray = ndarray[indices...]
    slice_expected::Array = expected[indices...]

    contentMatch = (slice_test == slice_expected)

    if !contentMatch
        println("Comparison failed on slice along with indices $indices")
        println("Ranges were $indices")
        println("Expected slice: ", slice_expected)
        println("Got slice:      ", slice_test)
    end
    return contentMatch
end


# ---------------------------------------------------------------------------------------------------------------------------------#
#                                                                                                                                  #
#                                                  Tests Utils ( check exceptions )                                                #
#                                                                                                                                  #
# ---------------------------------------------------------------------------------------------------------------------------------#

function exceptionPassed(f::Function, nb::Int, ExceptionType::Type)
    passed = true

    try
        x = f()
        passed = false
        println("      No exception thrown, in exception: $(nb)")
    catch e
        if (isa(e, ExceptionType))
            passed = true
        else
            passed = false
            println("      Wrong exception in exception: $(nb), excepted: $ExceptionType was: $(typeof(e))")
            rethrow(e)
        end
    end
    return passed
end

function allExceptionPassed(exceptions::Base.Vector{Tuple{Function,DataType}})
    for (i, tuple) in enumerate(exceptions)
        f, exceptionType = tuple[1], tuple[2]
        exceptionPassed(f, i, exceptionType) || return false
    end
    return true
end

# ---------------------------------------------------------------------------------------------------------------------------------#
#                                                                                                                                  #
#                                                  Tests Utils ( write to logs )                                                #
#                                                                                                                                  #
# ---------------------------------------------------------------------------------------------------------------------------------#
# Function to write timestamped messages to the log




function write_content(ndarray, write_log_fn::Function)
    logShape(ndarray, write_log_fn)
    logByIndex(ndarray, write_log_fn)
    logByIterator(ndarray, write_log_fn)
    logBySlices(ndarray, write_log_fn)
end


function write_content(scalar::AbstractFloat, write_log_fn::Function)
    write_log_fn(string(round(scalar, digits=7)))
end

function write_content(scalar::Number, write_log_fn::Function)
    write_log_fn(string(scalar))
end


function write_content(ndarray::NDArray{T, 0}, write_log_fn::Function) where {T}
    write_log_fn("$(ndarray[()])")
    logShape(ndarray, write_log_fn)
end

function write_content(ndarray::AbstractArray{T, 0}, write_log_fn::Function) where {T}
    write_log_fn("$(ndarray[])")
    logShape(ndarray, write_log_fn)
end

function write_content(ndarray::TestDataScalar, write_log_fn::Function)
    write_log_fn("$(ndarray.data)")
end


logShape(ndarray::AbstractArray, write_log_fn::Function) = logShape(ndims(ndarray), Base.prod(size(ndarray)), size(ndarray), write_log_fn)

logShape(ndarray::NDArray, write_log_fn::Function) = logShape(ndims(ndarray), ndarray.shape.length, size(ndarray), write_log_fn)

logShape(testData::ShapeData, write_log_fn::Function) = logShape(testData.dim, testData.total_element, testData.shape, write_log_fn)

function logShape(dim, total_element, shape, write_log_fn::Function)
    toLog = "dim: " * string(dim) * "\ntotal_element: " * string(total_element) * "\nshape: " * string(shape)
    write_log_fn(toLog)
end


logByIndex(testData::TestDataArray, write_log_fn::Function) = logByIndex(testData.matrix, write_log_fn)

function logByIndex(array::AbstractArray, write_log_fn::Function)
    ranges = ntuple(i -> 1:size(array)[i], ndims(array))
    contentByIndex = "index: "

    for indices in Iterators.product(ranges...)
        got = array[indices...]
        scalar =  typeof(got) <: AbstractFloat ? round(got, digits=7) : got
        contentByIndex *= "$(pretty_index(indices)): $scalar, "
    end
    write_log_fn(contentByIndex)
end


logByIterator(testData::TestDataArray, write_log_fn::Function) = logByIterator(testData.matrix, write_log_fn)

function logByIterator(array::AbstractArray, write_log_fn::Function)
    contentByIterator = "iterator: "
    for got in array
        s = typeof(got) <: AbstractFloat ? round(got, digits=7) : got
        contentByIterator *= "$s "
    end
    write_log_fn(contentByIterator)
end


logBySlices(testData::TestDataArray, write_log_fn::Function) = logBySlices(testData.matrix, write_log_fn)


function logBySlices(array::AbstractArray, write_log_fn::Function)
    shape = size(array)

    contentBySlices = ""

    write_log_fn("slices:")

    for axis in 1:ndims(array)
        for idx in CartesianIndices(ntuple(i -> i==axis ? 1 : 1:shape[i], ndims(array)))
        slice_idx = Tuple(idx)  # indices for the non-axis dimensions
        full_idx = ntuple(i -> i == axis ? Colon() : slice_idx[i], ndims(array))
        println("slices made: A[$(full_idx)]")
        contentBySlices *= logSlices(array, full_idx)
        end

    end
    write_log_fn(contentBySlices)
end



function round(x; digits::Int)
    if (x == -0.0)
        x = 0.0
    end
    return Base.round(x, digits=digits) 
end

function logSlices(matrix::AbstractArray{<:AbstractFloat}, indices)
    slice_test = matrix[indices...]          # get slice
    # convert each element to rounded number string
    slice_str = isempty(slice_test) ? "" : 
                "[" * join(round.(slice_test, digits=7), ", ") * "]"
    
    str = isempty(slice_test) ? 
          "array[$(pretty_index(indices...))] = []\n" :
          "array[$(pretty_index(indices...))] = $slice_str\n"
    
    return str
end

function logSlices(matrix::AbstractArray, indices)
    slice_test = matrix[indices...]
    str = isempty(slice_test) ? "array[$(pretty_index(indices...))] = []\n" : "array[$(pretty_index(indices...))] = $slice_test\n"
    return replace(str, "Any" => "")
end

end # TestUtils